lib/netplay/netqueue.cpp

```diff
@@ -106,78 +107,67 @@ bool decode_uint32_t(uint8_t b, uint32_t &v, unsigned n)
 	return !isLastByte;
 }
 
-uint8_t *NetMessage::rawDataDup() const
+NetMessage::NetMessage(std::vector<uint8_t>&& data)
+	: data_(std::move(data))
+{}
+
+uint8_t NetMessage::type() const
 {
-#if SIZE_MAX > UINT32_MAX
-	ASSERT_OR_RETURN(nullptr, data.size() <= static_cast<size_t>(std::numeric_limits<uint32_t>::max()), "Trying to duplicate a very large packet (%zu bytes). (Message type: %" PRIu8 ")", data.size(), type);
-#endif
-	uint32_t dataSizeU32 = static_cast<uint32_t>(data.size());
-	unsigned encodedLengthOfSize = encodedlength_uint32_t(dataSizeU32);
+	return data_[0];
+}
 
-	uint8_t *ret = new uint8_t[1 + encodedLengthOfSize + dataSizeU32];
+const std::vector<uint8_t>& NetMessage::rawData() const
+{
+	return data_;
+}
 
-	ret[0] = type;
-	uint32_t len = dataSizeU32;
-	for (unsigned n = 0; n < encodedLengthOfSize; ++n)
-	{
-		encode_uint32_t(ret[n + 1], len, n);
-	}
+const uint8_t* NetMessage::payload() const
+{
+	ASSERT_OR_RETURN(nullptr, !data_.empty() && data_.size() >= HEADER_LENGTH, "Invalid message data");
+	return &data_[HEADER_LENGTH];
+}
 
-	std::copy(data.begin(), data.begin() + dataSizeU32, ret + 1 + encodedLengthOfSize);
-	return ret;
+size_t NetMessage::payloadSize() const
+{
+	ASSERT_OR_RETURN(0, data_.size() >= HEADER_LENGTH, "Invlid message data");
+	return data_.size() - HEADER_LENGTH;
 }
 
-bool NetMessage::tryFromRawData(const uint8_t* buffer, size_t bufferLen, NetMessage& output)
+optional<NetMessage> NetMessage::tryFromRawData(const uint8_t* buffer, size_t bufferLen)
 {
-	if (bufferLen == 0) { return false; }
-	uint8_t type = buffer[0];
+	if (bufferLen == 0) { return nullopt; }
```

Comment:
Similarly to the `writeRawData` case, this probably needs to be:
```suggestion
	if (bufferLen < NetMessage::HEADER_LENGTH) { return nullopt; }
```

Otherwise there's a chance we could read beyond the end of `buffer` when trying to get `len` (if `bufferLen` < `NetMessage::HEADER_LENGTH`).